pkg/algorithm/caspool/v1alpha1/auto.go

```diff
@@ -0,0 +1,282 @@
+		WithAnnotations(op.CspcObject.GetAnnotations()).
+		WithDiskGroup(diskGroups).
+		WithCasTemplateName(op.CspcObject.GetCASTName()).
+		WithCspcName(op.CspcObject.Object.Name).
+		WithNodeName(nodeName).
+		WithDiskDeviceIDMap(diskDeviceIDMap).
+		Build().Object
+	return cp, nil
+}
+
+// validateAuto validates auto CSPC
+func (op *Operations) validateAuto() error {
+	err := NewOperationsBuilderForObject(op).
+		WithCheckf(IsMaxPoolNotNil(), "max pool is nil").
+		WithCheckf(IsPoolTypeValid(), "pool type is not valid").
+		WithCheckf(IsTypeValid(), "disk type is not valid").
+		Validate()
+	if err != nil {
+		return errors.Wrapf(err, "validation for cstorpoolcluster %s failed", op.CspcObject.Object.Name)
+	}
+	return nil
+}
+
+// getDisks returns a list of disk
+// The returned disk is
+// 1. Active
+// 2. Not used in other cstor-pool
+// 3. is not attached to a node where a cstor-pool has already been created for the given CSPC.
+func (op *Operations) getDisks() (*apisv1alpha1.DiskList, error) {
+	diskAPIList, err := op.DiskClient.List(metav1.ListOptions{})
+	if err != nil {
+		return nil, errors.Wrap(err, "could not list disk")
+	}
+
+	usedDiskMap, err := op.getUsedDiskMap()
+	if err != nil {
+		return nil, errors.Wrap(err, "could not get used disk map")
+	}
+
+	usedNodeMap, err := op.getUsedNode()
+	if err != nil {
+		return nil, errors.Wrap(err, "could not get used node map")
+	}
+
+	newDiskAPIList := disk.NewListBuilderForAPIList(diskAPIList).DiskList.
+		Filter(disk.IsActive(), disk.IsType(op.CspcObject.Object.Spec.Type), disk.IsUsable(usedDiskMap), disk.IsUsableNode(usedNodeMap)).
+		ObjectList
+
+	if len(newDiskAPIList.Items) == 0 {
+		return nil, errors.New("no disks found")
+	}
+
+	return newDiskAPIList, nil
+}
+
+// getNodeDisk returns a map of node and attached disk.
+// The returned node disk
+// 1. Obeys pool topology strictly with no spares
+// e.g.
+// for striped number of disk >1
+// for mirrored number of disk in a group = 2, number of group = k, total no. of disk - 2*k
+// Similarly, for other raid groups
+func (op *Operations) getNodeDisk() (map[string]*apisv1alpha1.DiskList, error) {
+	diskAPIList, err := op.getDisks()
+
+	if err != nil {
+		return nil, errors.Wrap(err, "could not get usable disk")
+	}
+	// TODO: Think of nodeDiskMap as a internal structure and apply creational and filter pattern
+	nodeDiskMap := make(map[string]*apisv1alpha1.DiskList)
+	for _, diskAPIObject := range diskAPIList.Items {
+		// pin it
+		diskAPIObject := diskAPIObject
+		nodeName := disk.BuilderForAPIObject(&diskAPIObject).Disk.GetNodeName()
+		if nodeDiskMap[nodeName] == nil {
+			nodeDiskMap[nodeName] = &apisv1alpha1.DiskList{Items: []apisv1alpha1.Disk{diskAPIObject}}
+		} else {
+			nodeDiskMap[nodeName].Items = append(nodeDiskMap[nodeName].Items, diskAPIObject)
+		}
+	}
+	filteredNodediskMap := filterNodeDiskMap(nodeDiskMap, isTopologyObeyed(nodeDiskMap, op.CspcObject.Object.Spec.PoolSpec.PoolType))
+
+	if len(filteredNodediskMap) == 0 {
```

Comment:
Can this check be moved to caller part.
This function behaviour is limited to finding disks based on pooltype.